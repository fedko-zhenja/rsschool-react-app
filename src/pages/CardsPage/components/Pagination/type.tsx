export interface PaginationProps {
    isDataLoaded: boolean;
    pageNumber: number;
    totalPages: number;
    onPageNumberChange: (value: string) => void;
}

export interface PaginationState {
    activePage: number;
    activeButtonNumber: Array<number>;
}
