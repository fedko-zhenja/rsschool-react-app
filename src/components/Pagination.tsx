import { useCallback } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { StoreState } from '../types/types';
import { setPageNumberValue } from '../lib/cardsReducer';
import styles from '../styles/Pagination.module.css';

const numberBtnOnPage = 5;

export function Pagination() {
    const dispatch = useDispatch();

    const reduxPageSizeValue = useSelector((state: StoreState) => state.cards.pageSizeValue);
    const reduxPageNumberValue = useSelector((state: StoreState) => state.cards.pageNumberValue);
    const reduxCardsData = useSelector((state: StoreState) => state.cards.cardsData);
    const reduxIsDataLoaded = useSelector((state: StoreState) => state.cards.isDataLoaded);

    const pageNumber = Number(reduxPageNumberValue);
    const totalPages = Math.ceil(reduxCardsData.totalCount / Number(reduxPageSizeValue));

    const handlePageChange = useCallback(
        (pageNumber: number) => {
            dispatch(setPageNumberValue(String(pageNumber)));
        },
        [dispatch]
    );

    const renderPageNumbers = () => {
        const pageNumbers = [];
        const startPage = Math.max(1, pageNumber - Math.floor(numberBtnOnPage / 2));
        const endPage = Math.min(totalPages, startPage + numberBtnOnPage - 1);

        for (let i = startPage; i <= endPage; i++) {
            pageNumbers.push(
                <button
                    key={i}
                    className={i === pageNumber ? styles.activeBtn : styles.paginationBtn}
                    onClick={() => handlePageChange(i)}
                    data-testid="pagination-btn"
                >
                    {i}
                </button>
            );
        }

        return pageNumbers;
    };

    if (reduxIsDataLoaded === false) {
        return null;
    }

    return (
        <div className={styles.paginationWrapper}>
            <button className={styles.paginationBtn} onClick={() => handlePageChange(1)} disabled={pageNumber === 1}>
                First
            </button>

            <button
                className={styles.paginationBtn}
                onClick={() => handlePageChange(Math.max(1, pageNumber - 1))}
                disabled={pageNumber === 1}
            >
                <div className={styles.prevBtn} />
            </button>

            {renderPageNumbers()}

            <button
                className={styles.paginationBtn}
                onClick={() => handlePageChange(Math.min(totalPages, pageNumber + 1))}
                disabled={pageNumber === totalPages}
            >
                <div className={styles.nextBtn} />
            </button>

            <button
                className={styles.paginationBtn}
                onClick={() => handlePageChange(totalPages)}
                disabled={pageNumber === totalPages}
            >
                Last
            </button>
        </div>
    );
}
