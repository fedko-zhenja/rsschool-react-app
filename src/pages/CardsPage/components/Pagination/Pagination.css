.pagination-wrapper {
    display: flex;
    gap: 5px;
    padding-bottom: 50px;
}

.pagination-btn {
    width: fit-content;
    height: 45px;
    color: black;
    background-color: white;
    border: 2px solid black;
    cursor: pointer;
    transition: 0.2s;
}

.pagination-btn:hover {
    border-color: #fbdb32 !important;
    background-color: #fbdb32;
}

.pagination-btn:active {
    background-color: #fbdb32;
}

.next-btn {
    width: 20px;
    height: 20px;
    background-image: url('../../../../assets/icon/next.png');
    background-size: 100%;
    background-repeat: no-repeat;
}

.prev-btn {
    width: 20px;
    height: 20px;
    background-image: url('../../../../assets/icon/next.png');
    background-size: 100%;
    background-repeat: no-repeat;
    transform: rotate(180deg);
}

.active-btn {
    background-color: #fbdb32;
    border-color: #fbdb32;
}
