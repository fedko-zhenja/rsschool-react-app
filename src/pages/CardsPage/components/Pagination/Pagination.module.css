.paginationWrapper {
    display: flex;
    gap: 5px;
    padding-bottom: 50px;
}

.paginationBtn {
    width: fit-content;
    height: 45px;
    color: black;
    background-color: white;
    border: 2px solid black;
    cursor: pointer;
    transition: 0.2s;
}

.paginationBtn:hover {
    border-color: #fbdb32 !important;
    background-color: #fbdb32;
}

.paginationBtn:active {
    background-color: #fbdb32;
}

.nextBtn {
    width: 20px;
    height: 20px;
    background-image: url('../../../../assets/icon/next.png');
    background-size: 100%;
    background-repeat: no-repeat;
}

.prevBtn {
    width: 20px;
    height: 20px;
    background-image: url('../../../../assets/icon/next.png');
    background-size: 100%;
    background-repeat: no-repeat;
    transform: rotate(180deg);
}

.activeBtn {
    background-color: #fbdb32;
    border-color: #fbdb32;

    width: fit-content;
    height: 45px;
    color: black;
    cursor: pointer;
    transition: 0.2s;
}
