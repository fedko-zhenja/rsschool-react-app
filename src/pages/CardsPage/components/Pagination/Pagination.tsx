import { useCallback } from 'react';
import { useCardsContext } from '../../../../context/context';
import './Pagination.css';

const numberBtnOnPage = 5;

export function Pagination() {
    const { isDataLoaded, pageNumberValue, cardsData, pageSizeValue, setPageNumberValue } = useCardsContext()!;

    const pageNumber = Number(pageNumberValue);
    const totalPages = Math.ceil(cardsData.totalCount / Number(pageSizeValue));

    const handlePageChange = useCallback(
        (pageNumber: number) => {
            setPageNumberValue(String(pageNumber));
        },
        [setPageNumberValue]
    );

    const renderPageNumbers = () => {
        const pageNumbers = [];
        const startPage = Math.max(1, pageNumber - Math.floor(numberBtnOnPage / 2));
        const endPage = Math.min(totalPages, startPage + numberBtnOnPage - 1);

        for (let i = startPage; i <= endPage; i++) {
            pageNumbers.push(
                <button
                    key={i}
                    className={i === pageNumber ? 'active-btn pagination-btn' : 'pagination-btn'}
                    onClick={() => handlePageChange(i)}
                    data-testid="pagination-btn"
                >
                    {i}
                </button>
            );
        }

        return pageNumbers;
    };

    if (isDataLoaded === false) {
        return null;
    }

    return (
        <div className="pagination-wrapper">
            <button className="pagination-btn" onClick={() => handlePageChange(1)} disabled={pageNumber === 1}>
                First
            </button>

            <button
                className="pagination-btn"
                onClick={() => handlePageChange(Math.max(1, pageNumber - 1))}
                disabled={pageNumber === 1}
            >
                <div className="prev-btn" />
            </button>

            {renderPageNumbers()}

            <button
                className="pagination-btn"
                onClick={() => handlePageChange(Math.min(totalPages, pageNumber + 1))}
                disabled={pageNumber === totalPages}
            >
                <div className="next-btn" />
            </button>

            <button
                className="pagination-btn"
                onClick={() => handlePageChange(totalPages)}
                disabled={pageNumber === totalPages}
            >
                Last
            </button>
        </div>
    );
}
