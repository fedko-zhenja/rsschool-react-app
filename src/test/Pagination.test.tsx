import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Pagination } from '../pages/CardsPage/components/Pagination/Pagination';
import { mockCardsData, mockCardsDataIsNotLoaded } from './mockData';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureMockStore([thunk]);
const store = mockStore({
    cards: {
        cardsData: mockCardsData.cardsData,
        isDataLoaded: true,
        pageSizeValue: '4',
        pageNumberValue: '2',
    },
});

const storeIsNotLoaded = mockStore({
    cards: {
        cardsData: mockCardsDataIsNotLoaded.cardsData,
        isDataLoaded: false,
        pageSizeValue: '0',
        pageNumberValue: '0',
    },
});

describe('Pagination', () => {
    test('should update URL query parameter when page changes', async () => {
        render(
            <MemoryRouter>
                <Provider store={store}>
                    <Pagination />
                </Provider>
            </MemoryRouter>
        );

        const paginationBtn = screen.getByText('1');
        expect(paginationBtn).toBeInstanceOf(HTMLElement);

        fireEvent.click(paginationBtn);

        expect(mockCardsData.pageNumberValue).toBe('1');
    });

    test('should not display buttons if data is not loaded', async () => {
        const { container } = render(
            <MemoryRouter>
                <Provider store={storeIsNotLoaded}>
                    <Pagination />
                </Provider>
            </MemoryRouter>
        );

        expect(container.firstChild).toBeNull();
    });
});
