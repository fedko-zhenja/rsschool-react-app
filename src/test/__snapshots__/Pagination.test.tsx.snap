// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination should do snapshot check 1`] = `
<DocumentFragment>
  <div
    class="pagination-wrapper"
  >
    <button
      class="pagination-btn"
      disabled=""
    >
      First
    </button>
    <button
      class="pagination-btn"
      disabled=""
    >
      <div
        class="prev-btn"
      />
    </button>
    <button
      class="active-btn pagination-btn"
      data-testid="pagination-btn"
    >
      1
    </button>
    <button
      class="pagination-btn"
      data-testid="pagination-btn"
    >
      2
    </button>
    <button
      class="pagination-btn"
      data-testid="pagination-btn"
    >
      3
    </button>
    <button
      class="pagination-btn"
    >
      <div
        class="next-btn"
      />
    </button>
    <button
      class="pagination-btn"
    >
      Last
    </button>
  </div>
</DocumentFragment>
`;
